### PR TITLE
Fix/wrong connection closure

### LINKED ISSUE #1893: SDL must close only session in case mobile app does not answer on Heartbeat_request
## Occurrence: 
Always 

## Steps to reproduce:
1. Reduce default value for Heartbeat timeout in SDL .INI file to 5 sec.
2. Start SDL
3. Start HMI
4. Connect mobile device
5. Register mobile application (sending of Heartbeat from application is switched off)
6. Wait 15 seconds
7. Check that app is unregistered
8. Check that mobile connection is still alive

## Expected result:
Mobile connection is still alive and it's possible to create new mobile session

## Actual result:
Mobile connection is closed

## Test script:
[1893_ATF_HeartBeat_App_does_not_send_HB_and_does_not_respond.lua](https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/develop/test_scripts/Defects/4_5/1893_ATF_HeartBeat_App_does_not_send_HB_and_does_not_respond.lua)

# Expected delivery
- [ ] Source code updates
- [ ] Code comments
- [x] **UTs add/update _(not required)_**
- [ ] ATF tests add/update
- [x] **Manual tests _(not required)_**
- [x] **Add/update CI plans/jobs _(not required)_**
- [x] **SDD updates _(not required)_**
- [x] **Guidelines update ([sdl_core_guides](https://github.com/smartdevicelink/sdl_core_guides)) _(not required)_**
- [x] **Guidelines update ([sdl_hmi_integration_guidelines](https://github.com/smartdevicelink/sdl_hmi_integration_guidelines)) _(not required)_**

### PR DESCRIPTION
**Pull request by AByzhynar**
Fixes #1893 

This PR is **[ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Testing Plan
Appropriate unit tests added/updated

### Summary
 - Removed connection closure in case of last session in the connection was closed.
   All connections should be removed when device would be unplugged.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/connection_handler/src/heartbeat_monitor.cc [hunk 1]
```diff
@@ -90,18 +90,18 @@ void HeartBeatMonitor::threadMain() {
 }
 
 void HeartBeatMonitor::AddSession(uint8_t session_id) {
-  LOG4CXX_DEBUG(logger_,
-                "Add session with id " << static_cast<int32_t>(session_id));
+  const uint32_t converted_session_id = static_cast<int32_t>(session_id);
```
Comment (LuxoftAKutsan) [inline src/components/connection_handler/src/heartbeat_monitor.cc]: @AByzhynar This variable used only in logger, Please mark it as UNUSED

File: src/components/connection_handler/test/connection_handler_impl_test.cc [hunk 2]
```diff
@@ -98,24 +98,26 @@ class ConnectionHandlerTest : public ::testing::Test {
...
-    ON_CALL(mock_connection_handler_settings, heart_beat_timeout())
+    ON_CALL(mock_connection_handler_settings_, heart_beat_timeout())
         .WillByDefault(Return(1000u));
     connection_handler_->addDeviceConnection(device_info, uid_);
     connection_key_ = connection_handler_->KeyFromPair(uid_, 0u);
     // Remove all specific services
   }
-  void AddTestSession() {
+
+  uint32_t AddTestSession() {
```
Comment (LuxoftAKutsan) [inline src/components/connection_handler/test/connection_handler_impl_test.cc]: @AByzhynar Provide description please

File: src/components/connection_handler/src/heartbeat_monitor.cc (hunk 1 above)
Comment (AByzhynar) [reply src/components/connection_handler/src/heartbeat_monitor.cc]: @LuxoftAKutsan Done in 8033d26b57aa342fcd191c58bcac5a46c00a649a

File: src/components/connection_handler/test/connection_handler_impl_test.cc (hunk 2 above)
Comment (AByzhynar) [reply src/components/connection_handler/test/connection_handler_impl_test.cc]: @LuxoftAKutsan Done in da7e745e88ac17a2efe416236222f1887549cbde

File: src/components/connection_handler/src/connection.cc [hunk 3]
```diff
@@ -310,24 +310,17 @@ const SessionMap Connection::session_map() const {
 }
 
 void Connection::CloseSession(uint8_t session_id) {
-  size_t size;
```
Comment (yang1070) [inline src/components/connection_handler/src/connection.cc]: this  change looks good to me

**Comment by AByzhynar** [comment]
@JackLivio Any details? What script you are running?

File: src/components/connection_handler/test/connection_handler_impl_test.cc [hunk 4]
```diff
@@ -376,23 +382,23 @@ TEST_F(ConnectionHandlerTest, GetAppIdOnSessionKey) {
...
 TEST_F(ConnectionHandlerTest, GetAppIdOnSessionKey_SessionNotStarted) {
   AddTestDeviceConnection();
 
   uint32_t app_id = 0;
-  connection_handler::DeviceHandle null_handle = 0;
+  connection_handler::DeviceHandle* device_id = NULL;
   EXPECT_EQ(-1,
             connection_handler_->GetDataOnSessionKey(
-                connection_key_, &app_id, NULL, &null_handle));
+                connection_key_, &app_id, NULL, device_id));
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/connection_handler/test/connection_handler_impl_test.cc]: @jacobkeeler No as any type of pointer can have `NULL` value. And as we have 2 functions:
```
int32_t ConnectionHandlerImpl::GetDataOnSessionKey(
    uint32_t key,
    uint32_t* app_id,
    std::list<int32_t>* sessions_list,
connection_handler::DeviceHandle* device_id) const
```
and
```
int32_t ConnectionHandlerImpl::GetDataOnSessionKey(
    uint32_t key,
    uint32_t* app_id,
    std::list<int32_t>* sessions_list,
uint32_t* device_id) const
```
it leads to ambiguity. The compiler does not know which function to call. Therefore we have to define explicitly type of this variable despite of its value will be `NULL`.

File: src/components/connection_handler/test/connection_handler_impl_test.cc [hunk 5]
```diff
@@ -37,10 +37,8 @@
 #include "connection_handler/connection_handler_impl.h"
 #include "protocol/common.h"
 // TODO(EZamakhov): move security test
-#ifdef ENABLE_SECURITY
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/connection_handler/test/connection_handler_impl_test.cc]: @jacobkeeler Right. I will return it back. Thanks.

[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/connection_handler/test/connection_handler_impl_test.cc]: @jacobkeeler Fixed in daed1660dfdc91cf8227b5637f86224eb1233e5e

File: src/components/connection_handler/test/connection_handler_impl_test.cc [hunk 6]
```diff
@@ -177,7 +183,7 @@ class ConnectionHandlerTest : public ::testing::Test {
     ASSERT_FALSE(connection_list.empty());
     ConnectionList::const_iterator conn_it = connection_list.find(connectionId);
     ASSERT_NE(conn_it, connection_list.end());
-    const Connection& connection = *conn_it->second;
+    const Connection& connection = *connection_list.begin()->second;
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/connection_handler/test/connection_handler_impl_test.cc]: @jacobkeeler Fixed in daed1660dfdc91cf8227b5637f86224eb1233e5e

File: src/components/connection_handler/test/connection_handler_impl_test.cc [hunk 7]
```diff
@@ -1978,41 +1977,6 @@ TEST_F(ConnectionHandlerTest, RunAppOnDevice_AppOnDevice_SUCCESS) {
...
-  connection_handler::Device d2(device_info_2.device_handle(),
-                                device_info_2.name(),
-                                device_info_2.mac_address(),
-                                device_info_2.connection_type());
-
-  EXPECT_CALL(mock_connection_handler_observer,
-              OnDeviceSwitchingStart(SameDevice(d1), SameDevice(d2)));
-
-  connection_handler_->OnDeviceSwitchingStart(mac_address_, second_mac_address);
-}
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/connection_handler/test/connection_handler_impl_test.cc]: @jacobkeeler Fixed in daed1660dfdc91cf8227b5637f86224eb1233e5e

**Pull request merged**